### ISSUE TITLE
Aria-invalid state description may need clarification for on blur validation

### ISSUE THREAD

**Issue by jorycunningham** [w3c/aria#1470]
The [state description for aria-invalid](https://www.w3.org/TR/wai-aria-1.2/#aria-invalid) says:

>  However, if the user has not attempted to submit the form, authors SHOULD NOT set the aria-invalid attribute on required widgets simply because the user has not yet entered data.

This seems to imply that using aria-invalid to a field based on an input-local event (e.g. on change, blur) is not an acceptable use of aria-invalid.

[WCAG 2.1 G83 ](https://www.w3.org/WAI/WCAG21/Techniques/general/G83.html) for error suggestion specifically identifies the a local on change event as a valid validation point. Specifically:

> A user is completing a form that contains mandatory fields. The labels of the fields indicate whether or not they are mandatory. The user tabs to a mandatory field, and tabs out of the field without entering any data or selecting a choice. A client-side script modifies the label of the field to indicate that leaving it blank was an error.

It would seem that using aria-invalid="true"  in this case would be a desirable way to to mirror the visual error state.

I would suggest either removing that line or changing the above quoted aria-invalid description line to read something like:

>  However, if the user has not attempted to submit the form or interacted with the field then moved away from it, authors SHOULD NOT set the aria-invalid attribute on required widgets simply because the user has not yet entered data.

**Comment by JAWS-test** [comment]
I wonder if the WCAG or the ARIA specification should be changed. On the one hand, validating fields after leaving is very common, on the other hand, I think it's annoying when I navigate through a form to explore it and I get an error message from the screen reader after every tab step, e.g. via live region.  It can also be annoying for cognitively impaired people when they explore a form and errors are continuously displayed. So I would rather have the question checked with WCAG first.

**Comment by jorycunningham** [comment]
I can appreciate your perspective on that @JAWS-test , but validation on blur is an extremely common practice. Might it make sense to normalize with current WCAG, trusting that that working group considered the potential drawbacks you described, and then change it if at such a time they decide to change this?

My concern is that folks will implement validation on blur without aria-invalid if they read the expressed prohibition in the state description, I would argue that would be a huge negative for the user.

**Comment by scottaohara** [comment]
it's not really prohibited though, it's a 'should not' not a 'must not'.

I would also agree with JAWS-test in getting this squared with WCAG.  That's not a normative document you referenced, and may very well need a hard look / update from the WG, especially since the quoted text of that doc is unchanged since the 2.0 version... and expected behaviors and markup patterns have changed since then.

**Comment by jorycunningham** [comment]
Okay, so we are comfortable with the ARIA spec saying that authors **should not** implement a common pattern, which is also WCAG-endorsed, using the tool for the job that is most appropriate: aria-invalid?

Not intending to be difficult or facetious with that question, I just want to make sure we are comfortable with the outcome of this choice.

**Comment by scottaohara** [comment]
@jorycunningham no choice has been made.  Merely clarifying _your_ assertion that ARIA is prohibiting this practice, which it is not. If ARIA was prohibiting this, it would be stated as such.

The suggestion to connect with WCAG about their techniques is due diligence.  There are _many_ WCAG techniques that need to be updated from time to time (take a peark at their GitHub issues... there are some heavy discussions going on about some techniques that _need_ updating).  Maybe this is one of them as well?  The text of that particular example even talks about modifying the label of the form control to indicate the invalid state – clearly that is not something you should do in conjunction with setting an `aria-invalid` state, as that would be a redundancy.

**Comment by jorycunningham** [comment]
Okay, I'll close this then as I do not at all disagree with WCAG on this and would not want to advocate for something I disagree with.

**Comment by JAWS-test** [comment]
@jorycunningham See https://github.com/w3c/wcag/issues/1769

**Comment by jorycunningham** [comment]
Thanks @JAWS-test .

I do not feel as strongly about throwing an error (and also aria-invalid="true") on a field left blank (as in your quite reasonable "I should be allowed to explore the page" example) but if you enter something demonstrably invalid and blur away, I think throwing an error is at least legitimate and arguably a UX imperative. I'd support clarifying WCAG there, but I still find it sub-optimal that ARIA is implicitly saying we should not validate forms other than as a side effect of hitting submit.

I say implicitly because I think saying that we should not use aria-invalid unless a submit attempt occurs is essentially the same as saying you should not validate the field at all unless there is a submit.

**Comment by JAWS-test** [comment]
@scottaohara Note, however, that some browsers (such as Firefox and Chrome) mark an empty required field (required attribute) or other deviations from the expected format as invalid even before focus is received. I.e. these fields are always automatically output as incorrect by the screen reader. Therefore my question would be, if for this reason the recommendation at aria-invalid should not be removed after all, to be in accordance with the standard HTML behavior. Or would you consider the current browser behavior as bugs?

**Comment by scottaohara** [comment]
Seems a bug to me. Doesn’t align with the default visual display which does not show the invalid styling until after a user has purposefully interacted with the field.

**Comment by JAWS-test** [comment]
@scottaohara I'm not sure, but this doesn't seem to be a bug of the browsers, but a requirement of the HTML specification, see https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-being-missing.

**Comment by scottaohara** [comment]
Hmm, seems so.  Then the default visual display for browsers doesn't match the fact that these are exposed as invalid by default.  So apparently browsers made a choice to expose a required field as invalid by default, but not visually convey it as such until after it is interacted with.
